# Data cleaning

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("clean_jobs.csv")
print(df)

        id                                              title  \
0        1                                       Data Analyst   
1        2                                       Data Analyst   
2        3                                       Data Analyst   
3        4                                       Data Analyst   
4        5                                    Data Analyst II   
...    ...                                                ...   
1043  2521                 Senior GIS/Data Management Analyst   
1044  2531  Data Science Senior Associate - Card Data & An...   
1045  2532  Vice President Data Scientist Marketing Analyt...   
1046  2533           Lead Data Scientist - Finance Technology   
1047  2534                     Applied AI/ML - Vice President   

                      company             location  \
0                        Meta         New York, NY   
1                        Meta    San Francisco, CA   
2                        Meta      Los Angeles, CA   
3  

In [3]:
df.shape

(1048, 10)

In [4]:
df.columns

Index(['id', 'title', 'company', 'location', 'link', 'source', 'date_posted',
       'work_type', 'employment_type', 'description'],
      dtype='object')

In [5]:
df.isnull().sum()

id                    0
title                 0
company               2
location             15
link                  0
source               89
date_posted          23
work_type          1048
employment_type    1048
description           4
dtype: int64

In [6]:
df = df.drop(columns=["work_type", "source", "employment_type", "link"])

In [7]:
print(df)

        id                                              title  \
0        1                                       Data Analyst   
1        2                                       Data Analyst   
2        3                                       Data Analyst   
3        4                                       Data Analyst   
4        5                                    Data Analyst II   
...    ...                                                ...   
1043  2521                 Senior GIS/Data Management Analyst   
1044  2531  Data Science Senior Associate - Card Data & An...   
1045  2532  Vice President Data Scientist Marketing Analyt...   
1046  2533           Lead Data Scientist - Finance Technology   
1047  2534                     Applied AI/ML - Vice President   

                      company             location date_posted  \
0                        Meta         New York, NY  2025-04-14   
1                        Meta    San Francisco, CA  2025-04-14   
2                    

In [8]:
df.isnull().sum()

id              0
title           0
company         2
location       15
date_posted    23
description     4
dtype: int64

# Handling missing values

In [9]:
df = df.dropna(subset=["description"])
df.isnull().sum()

id              0
title           0
company         2
location       15
date_posted    23
description     0
dtype: int64

In [10]:
df = df.dropna(subset=["company"])
df.isnull().sum()

id              0
title           0
company         0
location       15
date_posted    23
description     0
dtype: int64

In [11]:
df["location"] = df["location"].fillna("Unknown")
df.isnull().sum()

id              0
title           0
company         0
location        0
date_posted    23
description     0
dtype: int64

In [12]:
df["date_posted"] = df["date_posted"].fillna("Not Specified")
df.isnull().sum()

id             0
title          0
company        0
location       0
date_posted    0
description    0
dtype: int64

# Remove duplicates

In [13]:
df.duplicated(subset=["title", "company", "location"]).sum()

np.int64(55)

In [14]:
df = df.drop_duplicates(subset=["title", "company", "location"])
df.shape

(987, 6)

# Role category

In [15]:
df["title"] = df["title"].str.lower()
df["title"].head()

0       data analyst
1       data analyst
2       data analyst
3       data analyst
4    data analyst ii
Name: title, dtype: object

In [16]:
df["title"].value_counts().head(20)

title
data analyst                         149
data scientist                        73
data engineer                         54
machine learning engineer             21
data scientist, product analytics     15
data engineer, product analytics      14
senior data engineer                  13
junior data scientist                 12
senior data analyst                   11
ai/ml engineer                         9
**** *******                           9
junior data analyst                    8
business data analyst                  7
data analyst ii                        6
data analyst i                         6
**** *********                         6
**** ********, *********               6
**** ********                          6
data engineer i                        5
data products analyst, youtube         4
Name: count, dtype: int64

In [17]:
def categorize_role(title):
    if "data analyst" in title:
        return "Data Analyst"
    elif "business analyst" in title:
        return "Business Analyst"
    elif "product" in title and "analyst" in title:
        return "Product Analyst"
    elif "data scientist" in title:
        return "Data Scientist"
    elif "machine learning" in title or "ml engineer" in title or "ai/ml" in title:
        return "ML Engineer"
    elif "data engineer" in title:
        return "Data Engineer"
    else:
        return "Other"

df["role_category"] = df["title"].apply(categorize_role)

df["role_category"].value_counts()

role_category
Other               282
Data Analyst        267
Data Scientist      186
Data Engineer       185
ML Engineer          56
Product Analyst       6
Business Analyst      5
Name: count, dtype: int64

In [18]:
df[df["role_category"] == "Other"]["title"].value_counts().head(20)

title
**** *******                                           9
**** ********                                          6
**** ********, *********                               6
**** *********                                         6
analyst, analytics                                     4
analytics engineer                                     4
ai automation specialist                               4
business intelligence analyst                          3
data insights and analytics engineer                   2
******** **** *******                                  2
analytics engineer (l5) - ads                          2
data analytics                                         2
analista de business intelligence júnior               2
sr. analyst, data & analytics                          2
senior c# developer                                    2
devops engineer - $20k sign on bonus - 23285           2
department manager                                     2
multiple positions       

In [19]:
def categorize_role(title):
    if "data analyst" in title:
        return "Data Analyst"
    elif "business analyst" in title:
        return "Business Analyst"
    elif "business intelligence" in title or "bi analyst" in title:
        return "Business Analyst"
    elif "product" in title and "analyst" in title:
        return "Product Analyst"
    elif "analytics engineer" in title:
        return "Data Engineer"
    elif "data analytics" in title:
        return "Data Analyst"
    elif "analytics" in title and "analyst" in title:
        return "Data Analyst"
    elif "data scientist" in title:
        return "Data Scientist"
    elif "machine learning" in title or "ml engineer" in title or "ai/ml" in title:
        return "ML Engineer"
    elif "data engineer" in title:
        return "Data Engineer"
    else:
        return "Other"

df["role_category"] = df["title"].apply(categorize_role)

df["role_category"].value_counts()

role_category
Data Analyst        288
Other               237
Data Engineer       198
Data Scientist      186
ML Engineer          56
Business Analyst     16
Product Analyst       6
Name: count, dtype: int64

In [20]:
(df["role_category"].value_counts(normalize=True) * 100).round(2)

role_category
Data Analyst        29.18
Other               24.01
Data Engineer       20.06
Data Scientist      18.84
ML Engineer          5.67
Business Analyst     1.62
Product Analyst      0.61
Name: proportion, dtype: float64

# Skill extraction

In [21]:
df["description"] = df["description"].str.lower()
df["description"].head()

0    the social measurement team is a growing team ...
1    the social measurement team is a growing team ...
2    the social measurement team is a growing team ...
3    the social measurement team is a growing team ...
4    about pinterest\n\nmillions of people around t...
Name: description, dtype: object

In [22]:
df["python_required"] = df["description"].str.contains("python", na=False)
df["python_required"].sum()

np.int64(654)

In [23]:
df["sql_required"] = df["description"].str.contains("sql", na=False)
df["sql_required"].sum()

np.int64(707)

In [24]:
df["excel_required"] = df["description"].str.contains("excel", na=False)
df["excel_required"].sum()

np.int64(474)

In [25]:
df["powerbi_required"] = df["description"].str.contains("power bi", na=False)
df["powerbi_required"].sum()

np.int64(205)

In [26]:
df["tableau_required"] = df["description"].str.contains("tableau", na=False)
df["tableau_required"].sum()

np.int64(263)

In [27]:
df["ml_required"] = df["description"].str.contains("machine learning", na=False)
df["ml_required"].sum()

np.int64(347)

In [28]:
df["aws_required"] = df["description"].str.contains("aws", na=False)
df["aws_required"].sum()

np.int64(291)

In [29]:
df["communication_required"] = df["description"].str.contains("communication", na=False)
df["communication_required"].sum()

np.int64(520)

# Skill summary

In [30]:
skill_counts = {
    "Python": df["python_required"].sum(),
    "SQL": df["sql_required"].sum(),
    "Excel": df["excel_required"].sum(),
    "Power BI": df["powerbi_required"].sum(),
    "Tableau": df["tableau_required"].sum(),
    "Machine Learning": df["ml_required"].sum(),
    "AWS": df["aws_required"].sum(),
    "Communication": df["communication_required"].sum()
}

skill_df = pd.DataFrame(skill_counts.items(), columns=["Skill", "Count"])
skill_df.sort_values(by="Count", ascending=False)

,Skill,Count
1,SQL,707
0,Python,654
7,Communication,520
2,Excel,474
5,Machine Learning,347
6,AWS,291
4,Tableau,263
3,Power BI,205


### Preliminary Skill Demand Insights

- SQL appears to be the most frequently mentioned technical skill.
- Python closely follows SQL in demand.
- Communication skills are highly emphasized across roles.
- Traditional tools like Excel remain relevant.
- Machine Learning and AWS show strong demand in advanced roles.
- BI tools (Tableau, Power BI) are mentioned less frequently compared to core programming skills.

In [31]:
df.groupby("role_category")["sql_required"].mean().sort_values(ascending=False)

role_category
Data Engineer       0.888889
Product Analyst     0.833333
Data Analyst        0.826389
Data Scientist      0.736559
Business Analyst    0.625000
Other               0.506329
ML Engineer         0.375000
Name: sql_required, dtype: float64

In [32]:
df.groupby("role_category")["python_required"].mean().sort_values(ascending=False)

role_category
Data Scientist      0.881720
Data Engineer       0.863636
ML Engineer         0.839286
Product Analyst     0.666667
Other               0.506329
Data Analyst        0.500000
Business Analyst    0.250000
Name: python_required, dtype: float64

### Insight :

1. SQL is foundational across analyst and engineering roles.
2. Python is dominant in advanced roles (Data Scientist, ML Engineer).
3. Data Analyst roles rely significantly more on SQL than Python.
4. Business Analyst roles show minimal Python requirement.

In [33]:
df.groupby("role_category")["excel_required"].mean().sort_values(ascending=False)

role_category
Business Analyst    0.687500
Product Analyst     0.666667
Data Analyst        0.600694
Other               0.476793
ML Engineer         0.464286
Data Engineer       0.409091
Data Scientist      0.354839
Name: excel_required, dtype: float64

### Structure -

1. SQL → core for analysts & engineers
2. Python → core for scientists & ML roles
3. Excel → dominant in analyst/business roles

In [34]:
df.groupby("role_category")["ml_required"].mean().sort_values(ascending=False)

role_category
ML Engineer         0.767857
Data Scientist      0.672043
Product Analyst     0.666667
Data Engineer       0.338384
Other               0.278481
Data Analyst        0.142361
Business Analyst    0.062500
Name: ml_required, dtype: float64

In [35]:
df.groupby("role_category")["aws_required"].mean().sort_values(ascending=False)

role_category
Data Engineer       0.520202
ML Engineer         0.375000
Other               0.274262
Business Analyst    0.250000
Data Scientist      0.225806
Data Analyst        0.194444
Product Analyst     0.000000
Name: aws_required, dtype: float64

In [36]:
df.groupby("role_category")["powerbi_required"].mean().sort_values(ascending=False)

role_category
Data Analyst        0.402778
Other               0.181435
Product Analyst     0.166667
Business Analyst    0.125000
Data Scientist      0.112903
Data Engineer       0.106061
ML Engineer         0.017857
Name: powerbi_required, dtype: float64

In [37]:
df.groupby("role_category")["tableau_required"].mean().sort_values(ascending=False)

role_category
Data Analyst        0.493056
Business Analyst    0.375000
Data Scientist      0.220430
Data Engineer       0.171717
Product Analyst     0.166667
Other               0.164557
ML Engineer         0.000000
Name: tableau_required, dtype: float64

# Final Conclusion

This project analyzed 987 cleaned LinkedIn job postings across major data-related roles to identify hiring trends and skill demand patterns.

### 1. SQL Emerges as the Most Foundational Skill
SQL appears in the highest number of postings overall and dominates across Data Analyst (82.6%) and Data Engineer (88.9%) roles. This indicates that database proficiency is a core requirement across analytical and engineering positions.

### 2. Python Drives Advanced Technical Roles
Python demand is strongest in Data Scientist (88.1%) and ML Engineer (83.9%) roles, confirming its importance in modeling, automation, and advanced analytics workflows.

### 3. Data Analyst Roles Are Visualization-Heavy
Data Analyst postings show strong demand for Tableau (49.3%) and Power BI (40.3%), along with Excel (60%). This highlights the reporting and dashboard-oriented nature of analyst roles.

### 4. Machine Learning Is Role-Specific
Machine Learning is highly concentrated in ML Engineer (76.7%) and Data Scientist (67.2%) roles but minimal in Data Analyst (14.2%) and Business Analyst (6.2%) roles. This indicates clear specialization differences within data careers.

### 5. Cloud Skills Are Critical for Engineering Roles
AWS demand is highest in Data Engineer roles (52%), showing the growing importance of cloud infrastructure in modern data pipelines.

### 6. Soft Skills Matter
Communication is mentioned in over 50% of postings, demonstrating that technical proficiency alone is insufficient without strong stakeholder interaction skills.

---

### Overall Insight

The data reveals clear skill segmentation across roles:

- **Data Analyst → SQL + Excel + BI tools**
- **Data Scientist → Python + Machine Learning + SQL**
- **Data Engineer → SQL + Python + AWS**
- **ML Engineer → Python + ML + Cloud**
- **Business Analyst → Excel + SQL + Reporting tools**

These patterns reflect the structured specialization within modern data careers and provide practical guidance for skill prioritization based on role targeting.

In [39]:
df.to_csv("cleaned_linkedin_jobs.csv", index = False)